### ISSUE TITLE
Level3 segfaults for dimension = 0

### ISSUE THREAD

**Issue by elmar-peise** [flame/blis#97]
I observe that BLIS segfaults for dgemm and other level 3 kernels when one of the dimensions is 0.  This can be replicated with the test suite on both the master and compose branch and on both Haswell and SandyBridge without threading.

**Comment by jeffhammond** [comment]
Hopefully we can implement the check for this only in the BLAS interface, while merely documenting that the BLIS interface requires sane arguments.

**Comment by devinamatthews** [comment]
I vote for fixing BLIS too.

**Comment by elmar-peise** [comment]
Checks for empty matrices with corresponding code branches are already implemented in BLIS: e.g., https://github.com/flame/blis/blob/master/frame/3/gemm/bli_gemm_int.c#L59 and the following lines.

As the following stack trace shows, the segfault actually happens after the invocation of `bli_gemm_int` in the attempt to free the control tree:

```
#0  0x0000000000472ba8 in bli_cntl_free (cntl=0x7c39b0, thread=0x0) at frame/base/bli_cntl.c:108
#1  0x0000000000472bc3 in bli_cntl_free (cntl=0x7c3a10, thread=0x7c3a70) at frame/base/bli_cntl.c:112
#2  0x000000000055dbc7 in bli_gemm_cntl_free (cntl=0x7c3a10, thread=0x7c3a70) at frame/3/gemm/bli_gemm_cntl.c:127
#3  0x000000000055d48a in bli_l3_cntl_free_if (a=0x7fffffff7730, b=0x7fffffff7690, c=0x7fffffff75f0, cntx=0x7fffffff78d0,
    cntl_orig=0x0, cntl_use=0x7c3a10, thread=0x7c3a70) at frame/3/bli_l3_cntl.c:100
#4  0x0000000000492eeb in bli_l3_thread_decorator (func=0x55e488 <bli_gemm_int>, alpha=0x7fffffff8540, a=0x7fffffff7730,
    b=0x7fffffff7690, beta=0x7fffffff8360, c=0x7fffffff75f0, cntx=0x7fffffff78d0, cntl=0x0) at frame/thread/bli_thrcomm_single.c:115
#5  0x000000000055e482 in bli_gemm_front (alpha=0x7fffffff8540, a=0x7fffffff84a0, b=0x7fffffff8400, beta=0x7fffffff8360,
    c=0x7fffffff82c0, cntx=0x7fffffff78d0, cntl=0x0) at frame/3/gemm/bli_gemm_front.c:99
#6  0x00000000004816a7 in bli_gemmnat (alpha=0x7fffffff8540, a=0x7fffffff84a0, b=0x7fffffff8400, beta=0x7fffffff8360,
    c=0x7fffffff82c0, cntx=0x0) at frame/ind/oapi/bli_l3_nat_oapi.c:74
#7  0x00000000004812c7 in bli_gemmind (alpha=0x7fffffff8540, a=0x7fffffff84a0, b=0x7fffffff8400, beta=0x7fffffff8360,
    c=0x7fffffff82c0, cntx=0x0) at frame/ind/oapi/bli_l3_ind_oapi.c:59
#8  0x000000000046fd20 in bli_gemm (alpha=0x7fffffff8540, a=0x7fffffff84a0, b=0x7fffffff8400, beta=0x7fffffff8360, c=0x7fffffff82c0)
    at frame/3/bli_l3_oapi.c:74
```

**Comment by fgvanzee** [comment]
@elmar-peise Just to be clear, this segfault has nothing to do with the level-3 kernels. But thank you for the report nonetheless.
Everyone else: This issue, which is squarely in the level-3 implementation, should be fixed in 22377ab. The fix is inherited across all user-level interfaces (BLIS, BLAS, CBLAS).

**Comment by elmar-peise** [comment]
Yes, that's what I thought but I did not have the problem for some level-1 and -2 kernels that I tried.
Thanks for the quick fix!
